Infrastructure: centralise predefined MUD game data

SlySven opened the pull request:
This moves the "description" data into the same container as all the other details for the predefined MUDS. That data is then refactored out of the `./src/mudlet.h` file into a separate smaller file that only needs to be `#include`d into two class files - so changing any details requires less to rebuild the application.

This PR also removes the 3rd party dependency "qt-ordered-map" (which is 3-clause BSD licensed but NOT currently mentioned in our "About" dialogue!) Afterwards the details for each predefined MUD is held in a `QList<T>` instead - with a couple of helper methods defined to reproduce the information retrieval methods used in the prior code.

This PR also adds an icon for the Mudlet self-test profile which eliminates a warning in development versions at least about a missing icon for that "game" - as such that particular profile is NOT select-able with a mouse and scrolling through the list with the arrow keys causes an adjacent profile to be highlighted in error!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Infrastructure: centralise predefined MUD game data

Moves the "description" data into the same container as all the other
details for the predefined MUDS. That data is then refactored out of the
`./src/mudlet.h` file into a separate smaller file that only needs to be
`#include`d into two class files - so changing any details requires less
to rebuild the application.

This PR also removes the 3rd party dependency "qt-ordered-map" (which is
3-clause BSD licensed but NOT currently mentioned in our "About" dialogue!)
Afterwards the details for each predefined MUD is held in a QList<T>
instead - with a couple of helper methods defined to reproduce the
information retrieval methods used in the prior code.

This PR also adds an icon for the Mudlet self-test profile which eliminates
a warning in development versions at least about a missing icon for that
"game" - as such that particular profile is NOT select-able with a mouse
and scrolling through the list with the arrow keys causes an adjacent
profile to be highlighted in error!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
The part about not having an icon declared in the details:

Current development branch - the selection sticks on one of the adjacent profiles:
![Screenshot_20221121_030925](https://user-images.githubusercontent.com/6163092/202955841-610150fb-21d8-4e5d-b1e4-43af50ac9f76.png)

and there is an error message on the OS command line:
![Screenshot_20221121_024417](https://user-images.githubusercontent.com/6163092/202955847-ce94bb3e-e40c-41a7-b753-1e48a71e788f.png)

After this PR which uses the otherwise unused `:/icons/generic_mug.jpg` icon:
![Screenshot_20221121_030438](https://user-images.githubusercontent.com/6163092/202955843-72c42115-9d2e-4867-858a-23f8f479ce97.png)

vadi2 reviewed (commented):
Awesome idea to move the game details out to their own file, `mudlet.h` is getting too big and full of unrelated things. Great that game order is preserved as well.

I don't think the "Mudlet self-test" profile is ready to be exposed yet though, it was intentional it didn't have an icon - mostly because the proper documentation isn't yet available for the players both within the profile and in the manual. We shouldn't be showing a new thing to people that they'll actually have not much clue to use, that would be dissapointing to them.

SlySven commented:
> I don't think the "Mudlet self-test" profile is ready to be exposed yet though, it was intentional it didn't have an icon - mostly because the proper documentation isn't yet available for the players both within the profile and in the manual. We shouldn't be showing a new thing to people that they'll actually have not much clue to use, that would be dissapointing to them.

Well it is present and will show up if the selection is moved around by using the keyboard - and thus will also be ~~seen~~ *noticed* by VIP users who are using the current PT Builds as I understand it. If it was intentional NOT to show it then further code to NOT create a `QListWidgetItem` for it is required...

vadi2 commented:
It is fine as it is right now. Indeed, it is accessible if you look for it and that is OK.

SlySven commented:
> It is fine as it is right now. Indeed, it is accessible if you look for it and that is OK.

Sorry, but I'm not clear, is the change I made to show it acceptable (leave PR as it is) or do you want to go back to how the prior code is (moaning about a missing icon and arrow keys behaving weirdly when the "selection" cursor passes over where it should be)...

vadi2 commented on src/TGameDetails.h:
```diff
@@ -0,0 +1,542 @@
+             23,
+             false,
+             qsl("<a href='https://www.mudlet.org'>www.mudlet.org"),
+             qsl(":/icons/generic_mud.jpg"),
```
```suggestion
             QString(), // intentionally not (very) visible yet as it needs to be documented first
```

SlySven replied:
As I say though - that is going to make the icon generator put out that error message and it messes with the navigation with arrow keys which ought to be noticed by those with concerns for **a11y**.

vadi2 replied:
I know. But this has been the case for a long while and has not been an issue. Observe how nobody complained or noticed.

We should only make this available when it's polished and documented.

vadi2 reviewed (approved):
This would do the trick

SlySven commented on src/TGameDetails.h:
```diff
@@ -0,0 +1,542 @@
+             23,
+             false,
+             qsl("<a href='https://www.mudlet.org'>www.mudlet.org"),
+             qsl(":/icons/generic_mud.jpg"),
```
```suggestion
             QString(), // TODO: https://github.com/Mudlet/Mudlet/issues/6443
```

SlySven committed:
Update src/TGameDetails.h

Partial reversion - of "fix" for missing Mudlet self-test icon.

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Infrastructure_centralisePredefinedMUDGameData

vadi2 committed:
Merge branch 'development' into Infrastructure_centralisePredefinedMUDGameData
